File: src/main/java/org/calvin/Book/pie/4-linked-lists/LinkedList.java
public class LinkedList {

    /**
     * Inset a node in the front.
     * Note: keeping tracking the head.
     * The caller should updates its reference to the new head element from the method.
     *  eg: head = insertInFront(head, data);
     */
    public ListElement<Integer> insertInFront(ListElement<Integer> list, int data) {
        ListElement<Integer> l = new ListElement<Integer>(data);
        l.setNext(list);
        return l;
    }

    /**
     * Travering a List: find a value in the list.
     * Note: If not found, it will return null.
     * The caller must detect an error condition by checking for a null return value.
     */
    public ListElement<Integer> find(ListElement<Integer> head, int data) {
        ListElement<Integer> elem = head;
        while (elem != null && elem.value() != data) {
            elem = elem.next();
        }
        return elem;
    }
}

File: src/main/java/org/calvin/Book/pie/4-linked-lists/LinkedList.c
#include <stdlib.h>     // malloc, free
#include <stdbool.h>    // bool
#include <ListElement.h>// ListElement

/**
 * Insert a node the front.
 * Note: We should pass a pointer to the head pointer to update the head.
 * Use return value to indicate the success or failure of the memory allocation(because there is no exception in C).
 */
bool insertInFront(ListElement **head, int data) {
    ListElement *newElem = malloc(sizeof(ListElement));
    if(!newElem) return false;

    newElem->data = data;
    newElem->next = *head;
    *head = newElem; // update the head pointer
    return true;
}

/**
 * Delete an element from the list.
 * Note: deletion require a pointer or reference to the element immediately preceding the deletion or insertion localtion.
 */
bool deleteElement(ListElement **head, ListElement *deleteMe) {
    ListElement *elem; /* previous node */

    if (!head || !*head || !deleteMe) /* check for null pointers */
        return false;

    elem = *head;
    if (deleteMe == *head) { /* special case for head */
        *head = elem->next;
        free(deleteMe);
        return true;
    }

    while (elem) {
        if (elem->next == deleteMe) {
            /* elem is element preceding deleteMe */
            elem->next = deleteMe->next;
            free(deleteMe);
            return true;
        }
        elem = elem->next;
    }
    /* deleteMe not found */
    return false;
}

/**
 * Remove all elements from list.
 * Note: using two pointers to solve free problem.
 * Deletion of an element always requires at least two pointers variables.
 */
void deleteList(ListElement **head) {
    ListElement *deleteMe = *head;
    while (deleteMe) {
        ListElement *next = deleteMe->next;
        free(deleteMe);
        deleteMe = next;
    }
    *head = NULL;
}
